src/client/src/producer.rs

```diff
@@ -96,82 +148,354 @@ impl TopicProducer {
+        Ok(())
+    }
+}
+
+async fn group_by_spu(
+    topic: &str,
+    partitions: &StoreContext<PartitionSpec>,
+    records_by_partition: Vec<(i32, DefaultRecord)>,
+) -> Result<HashMap<SpuId, HashMap<i32, Vec<DefaultRecord>>>, FluvioError> {
+    let mut map: HashMap<SpuId, HashMap<i32, Vec<DefaultRecord>>> = HashMap::new();
+    for (partition, record) in records_by_partition {
+        let replica_key = ReplicaKey::new(topic, partition);
+        let partition_spec = partitions
+            .lookup_by_key(&replica_key)
+            .await?
+            .ok_or_else(|| FluvioError::PartitionNotFound(topic.to_string(), partition))?
+            .spec;
+        let leader = partition_spec.leader;
+        map.entry(leader)
+            .or_insert_with(HashMap::new)
+            .entry(partition)
+            .or_insert_with(Vec::new)
+            .push(record);
+    }
 
-        let spu_client = self.pool.create_serial_socket(&replica).await?;
+    Ok(map)
+}
 
-        debug!("connect to replica leader at: {}", spu_client);
+fn assemble_requests(
+    topic: &str,
+    partitions_by_spu: HashMap<i32, HashMap<i32, Vec<DefaultRecord>>>,
+) -> Vec<(SpuId, DefaultProduceRequest)> {
+    let mut requests: Vec<(SpuId, DefaultProduceRequest)> =
+        Vec::with_capacity(partitions_by_spu.len());
 
-        let records = vec![(None, Bytes::from(Vec::from(buffer.as_ref())))];
-        send_records_raw(spu_client, &replica, records).await?;
-        Ok(())
+    for (leader, partitions) in partitions_by_spu {
+        let mut request = DefaultProduceRequest::default();
+
+        let mut topic_request = DefaultTopicRequest::default();
+        topic_request.name = topic.to_string();
+
+        for (partition, records) in partitions {
+            let mut partition_request = DefaultPartitionRequest::default();
+            partition_request.partition_index = partition;
+            partition_request
+                .records
+                .batches
+                .push(DefaultBatch::new(records));
+            topic_request.partitions.push(partition_request);
+        }
+
+        request.acks = 1;
+        request.timeout_ms = 1500;
+        request.topics.push(topic_request);
+        requests.push((leader, request));
+    }
+
+    requests
+}
+
+/// A trait for defining a partitioning strategy for key/value records.
+///
+/// A Partitioner is given a slice of potential keys, and the number of
+/// partitions in the current Topic. It must map each key from the input
+/// slice into a partition stored at the same index in the output Vec.
+///
+/// It is up to the implementor to decide how the keys get mapped to
+/// partitions. This includes deciding what partition to assign to records
+/// with no keys (represented by `None` values in the keys slice).
+///
+/// See [`SiphashRoundRobinPartitioner`] for a reference implementation.
+trait Partitioner {
+    fn partition(&mut self, key: Option<&[u8]>, value: &[u8]) -> i32;
```

Comment:
change i32 to SpuId

Reply:
I don't think this is a `SpuId`, this is the Partition. Is there a `PartitionId` type somewhere?

Reply:
There is `fluvio_types::PartitionId` which is `i32`, I can use that